### PR TITLE
Feature/api

### PR DESCRIPTION
**Pull request by JainishaParmar**
created APIs

### CODE REVIEW

**Review by jaimin45 — APPROVED** [review]

File: .eslintrc.json [hunk 1]
```diff
@@ -11,7 +11,8 @@
   "plugins": ["prettier", "spellcheck"],
   "rules": {
     "quotes": "off",
-    "prettier/prettier": "error",
+    "new-cap":"off",
```
Comment (kapilbarad) [inline .eslintrc.json]: We should not disable any rules. Remove this newly added rules.

File: app.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+const TutorialRouter = require('./routes/tutorial.routes');
+
+const dirname = path.resolve();
+
+const app = express();
+const swaggerDoc = swaggerYaml.load('./swagger.yaml');
+
+// eslint-disable-next-line no-console
+promise.then(() => console.log("Connection successfull!"));
+// eslint-disable-next-line no-console
```
Comment (kapilbarad) [inline app.js]: No need of this line here `// eslint-disable-next-line no-console`.

File: controllers/tutorial.controller.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
+/* eslint-disable no-console */
+const { response } = require('express');
+const { Tutorial } = require('../models/tutorial.schema');
+const mangoose = require('../init/db');
+
+// create Document or insert
+const createTutorial = async (req, res) => {
+  try {
```
Comment (kapilbarad) [inline controllers/tutorial.controller.js:5]: We should return a proper success or error response based on Swagger API Design here.

File: controllers/tutorial.controller.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    const demoTutorial = new Tutorial({
+      id: "1",
+      published: true,
+    });
+    const result = await demoTutorial.save();
+    // eslint-disable-next-line no-console
+    console.log(result);
```
Comment (kapilbarad) [inline controllers/tutorial.controller.js]: We should return newly created tutorial in response.

File: controllers/tutorial.controller.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  } catch (err) {
+    // eslint-disable-next-line no-console
+    console.log(err);
```
Comment (kapilbarad) [inline controllers/tutorial.controller.js]: We should return a proper error response.

File: controllers/tutorial.controller.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  }
+};
+
+const getTutorialById = async (req, res) => {
+  const findId = req.params.id;
+  try {
+    const tutorials = await Tutorial.find({ id: findId });
+    res.json(tutorials);
```
Comment (kapilbarad) [inline controllers/tutorial.controller.js]: 1. This API should only return a single Tutorial.
2. If the given tutorial ID doesn't exist in the database then it should return a proper response with `404` response code.

File: controllers/tutorial.controller.js [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  }
+};
+
+const getTutorial = async (req, res) => {
+  try {
+    const tutorials = await Tutorial.find();
+    res.json(tutorials);
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.log("error");
```
Comment (kapilbarad) [inline controllers/tutorial.controller.js]: We should return a proper error response from here.

File: controllers/tutorial.controller.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+  }
+};
+
+const getUpdateTutorial = async (data) => {
+  try {
+    const updateData = await Tutorial.update({ id: data }, { $set: { id: "2" } }).exec();
+    console.log(updateData);
```
Comment (kapilbarad) [inline controllers/tutorial.controller.js]: We should return proper response here base don API design.

File: controllers/tutorial.controller.js [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+  } catch (error) {
+    console.log("error", error);
+  }
+};
+
+const updateTutorial = async (req, res) => {
+  const data = req.params.id;
+  const tutorial = await getUpdateTutorial(data);
```
Comment (kapilbarad) [inline controllers/tutorial.controller.js]: We should just update the tutorial here , no need to define another function for that.

File: models/tutorial.schema.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
+const mangoose = require('mongoose');
+
+const tutorialSchema = new mangoose.Schema({
+  id: {
```
Comment (kapilbarad) [inline models/tutorial.schema.js]: ID is autogenerated in  mongoose no need to define it here.

File: models/tutorial.schema.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+    type: String,
+    required: true,
+  },
+  published: {
+    type: Boolean,
+    required: true,
+  },
+  createdAt: {
```
Comment (kapilbarad) [inline models/tutorial.schema.js]: `createdAt` and `updatedAt` can be autogenerated using some options. Please follow this docs - https://masteringjs.io/tutorials/mongoose/timestamps

File: routes/tutorial.routes.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
+const express = require('express');
+
+const router = express.Router();
+const {
+  createTutorial, getTutorial, updateTutorial, deleteTutorial, getTutorialById,
+} = require('../controllers/tutorial.controller');
+
+router.post("/create", createTutorial);
+router.get("/getTutorial", getTutorial);
```
Comment (kapilbarad) [inline routes/tutorial.routes.js]: Please use same paths for APIs as specified in the swagger.

**Review by kapilbarad — CHANGES_REQUESTED** [review]
@jaimin45 @ashishkava Please apply the same suggested changes in your code.

**Comment by kapilbarad** [comment]
@JainishaParmar In future, Please pull latest in develop before starting any new feature. Otherwise you'll see conflicts in code.

**Review by ghost — APPROVED** [review]

**Review by jaimin45 — APPROVED** [review]

File: models/tutorial.schema.js [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+// Schema
+const tutorialSchema = new mongoose.Schema(
+  {
+    published: {
+      type: Boolean,
+      required: true,
+    },
+  },
+  // eslint-disable-next-line spellcheck/spell-checker
```
Comment (kapilbarad) [inline models/tutorial.schema.js]: Please shift this to spellchecker ignore file.

File: models/tutorial.schema.js [hunk 14]
```diff
@@ -0,0 +1,15 @@
+const mongoose = require("mongoose");
+
+// Schema
+const tutorialSchema = new mongoose.Schema(
+  {
+    published: {
```
Comment (kapilbarad) [inline models/tutorial.schema.js:14]: Missing title, description fields in schema and swagger

File: routes/tutorial.routes.js [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+  updateTutorials,
+  deleteTutorials,
+} = require("../controllers/tutorial.controller");
+
+/* GET users listing. */
+router.get("/", tutorialsAll);
+router.get("/:tutorialId", tutorialSearchById);
+router.post("/", tutorialsCreate);
+router.put("/:tutorialId", updateTutorials);
+router.delete("/:tutorialId", deleteTutorials);
```
Comment (kapilbarad) [inline routes/tutorial.routes.js]: We should follow proper naming conventions. The `tutorialId` can be just `id`.

File: controllers/tutorial.controller.js [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    const { options } = { new: true };
+    const tutorialUpdate = await Tutorial.findByIdAndUpdate(
+      { _id: req.params.tutorialId },
+      { id: req.body.id },
+      update,
+      options,
+    );
+    res.json(tutorialUpdate);
+  } catch (error) {
+    res.status(400).send(error);
```
Comment (kapilbarad) [inline controllers/tutorial.controller.js]: - Log the error
- Should be 500 response with `{ message: 'Internal Server Error' }`.

File: controllers/tutorial.controller.js [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+};
+
+// Add new Tutorials
+const tutorialsCreate = async (req, res) => {
+  const tutorial = new Tutorial({
+    published: req.body.published,
+  });
+  try {
+    const savedTutorial = await tutorial.save();
+    res.send(savedTutorial);
```
Comment (kapilbarad) [inline controllers/tutorial.controller.js]: You can send `tutorial`.

File: controllers/tutorial.controller.js [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+  } catch (error) {
+    res.status(400).send(error);
+  }
+};
+
+// Update a Tutorial by ID
+const updateTutorials = async (req, res) => {
+  try {
```
Comment (kapilbarad) [inline controllers/tutorial.controller.js:32]:
```
const { title, description } = req.body
const update = {}
if (typeof title !== 'undefined') update.title = title
if (typeof description !== 'undefined') update.description = description
```

File: controllers/tutorial.controller.js [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+    const { update } = req.body;
+    const { options } = { new: true };
+    const tutorialUpdate = await Tutorial.findByIdAndUpdate(
+      { _id: req.params.tutorialId },
+      { id: req.body.id },
+      update,
+      options,
+    );
+    res.json(tutorialUpdate);
```
Comment (kapilbarad) [inline controllers/tutorial.controller.js]: Patch request should always return `204` - No Content.
```
res.status(204).send()
```

File: controllers/tutorial.controller.js [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+    res.json(tutorialRemove);
+  } catch (error) {
+    res.status(400).send(error);
+  }
+};
+
+// Able to search by Id
+const tutorialSearchById = async (req, res) => {
+  try {
+    const tutorials = await Tutorial.findById(req.params.tutorialId);
```
Comment (kapilbarad) [inline controllers/tutorial.controller.js]: Should be just `tutorial = `

File: controllers/tutorial.controller.js [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+    res.json(tutorials);
```
Comment (kapilbarad) [inline controllers/tutorial.controller.js]: Should return 404 when tutorial doesn't exist.
```
if (!tutorial) return res.status(404).send({message: 'Tutorial not found'})
```

File: controllers/tutorial.controller.js [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+  }
+};
+
+// Delete a Tutorial by ID
+const deleteTutorials = async (req, res) => {
+  try {
+    const tutorialRemove = await Tutorial.findByIdAndDelete(
+      req.params.tutorialId,
+    );
+    res.json(tutorialRemove);
```
Comment (kapilbarad) [inline controllers/tutorial.controller.js]: No need to return anything on delete.
```
res.status(204).send()
```

**Review by kapilbarad — CHANGES_REQUESTED** [review]
1. path in swagger should be `/tutorials`
2. Add missing api in swagger `GET /tutorials/:id`
3. Add missing responses of the APIs in swaager
```
GET /tutorials/:id
   200: When tutorial exists
   404: When tutorial doesn't exist
   
```

File: controllers/tutorial.controller.js [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+  } catch (error) {
+    res.status(500).send("Internal Server Error");
+  }
+};
+
+// Delete a Tutorial by ID
+const deleteTutorial = async (req, res) => {
+  try {
+    const tutorialRemove = await Tutorial.findByIdAndDelete(req.params.Id);
+    res.json(tutorialRemove);
```
Comment (kapilbarad) [inline controllers/tutorial.controller.js]: We're sending response two times here. Plz correct this.

File: routes/tutorial.routes.js [hunk 24]
```diff
@@ -0,0 +1,15 @@
+const router = require("express").Router();
+const {
+  getTutorials,
+  getTutorialById,
+  createTutorial,
+  updateTutorial,
+  deleteTutorial,
+} = require("../controllers/tutorial.controller");
+/* GET users listing. */
```
Comment (kapilbarad) [inline routes/tutorial.routes.js]: Please remove this unnecessary comment.

File: routes/tutorial.routes.js [hunk 25]
```diff
@@ -0,0 +1,15 @@
...
+router.get("/", getTutorials);
+router.get("/:Id", getTutorialById);
```
Comment (kapilbarad) [inline routes/tutorial.routes.js]: `Id` should be `id`

File: swagger.yaml [hunk 26]
```diff
@@ -15,7 +15,7 @@ consumes:
 produces:
   - application/json
 paths:
-  tutorials:
+  /tutorials:
     parameters:
       - name: id
```
Comment (kapilbarad) [inline swagger.yaml]: No need of parameters here as we're not using it in PATH anywhere.

File: swagger.yaml [hunk 27]
```diff
@@ -49,30 +52,51 @@ paths:
...
+        '400':
+          description: 'Invalid ID tutorial'
+
     patch:
       summary: Update tutorial with give ID
       tags:
         - Tutorials
       parameters:
-        - name: Tutorial
+        - name: title
```
Comment (kapilbarad) [inline swagger.yaml]: Why title? Please test and correct this. You can test it from swagger UI and see how it's passing data to body.

File: swagger.yaml [hunk 28]
```diff
@@ -81,6 +105,10 @@ definitions:
         type: string
         format: uuid
         uniqueItems: true
+      title:
```
Comment (kapilbarad) [inline swagger.yaml:131]: Please specify required fields.

**Review by kapilbarad — CHANGES_REQUESTED** [review]

**Review by jaimin45 — APPROVED** [review]

**Review by ghost — APPROVED** [review]

File: models/tutorial.schema.js [hunk 29]
```diff
@@ -0,0 +1,23 @@
...
+    description: {
+      type: String,
+      required: true,
+    },
+    published: {
+      type: Boolean,
+      required: true,
+    },
+  },
+  // eslint-disable-next-line spellcheck/spell-checker
```
Comment (jaimin45) [inline models/tutorial.schema.js]: remove this spell check line

**Review by jaimin45 — CHANGES_REQUESTED** [review]

**Review by jaimin45 — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Review by kapilbarad — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Review by jaimin45 — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Comment by JainishaParmar** [comment]
feature/api

**Pull request merged**